File: python/infinicore/ops/syr.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def syr(
    alpha: Tensor,
    x: Tensor,
    out: Tensor,
    *,
    uplo: int = 0,
):
    _infinicore.syr_(
        alpha._underlying,
        x._underlying,
        out._underlying,
        uplo,
    )

    return out

File: src/infinicore/ops/syr/syr.cc
#include "infinicore/ops/syr.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Syr);

Syr::Syr(const Tensor &alpha, const Tensor &x, Tensor a, int uplo) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(alpha, x, a);
    INFINICORE_GRAPH_OP_DISPATCH(a->device().getType(), alpha, x, a, uplo);
}

void Syr::execute(const Tensor &alpha, const Tensor &x, Tensor a, int uplo) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Syr, alpha, x, a, uplo);
}

void syr_(const Tensor &alpha, const Tensor &x, Tensor a, int uplo) {
    Syr::execute(alpha, x, a, uplo);
}

} // namespace infinicore::op
